### PR TITLE
Round primary color in swrast

### PR DESCRIPTION
**Pull request by ds84182**
Also includes a fix for logic ops not being enabled in OpenGL (depending on circumstances).

Also ports the hack from the hardware renderer to software rasterizer where null texture addresses are ignored. (Otherwise the software rasterizer will crash doing a null deref).

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3097)

### CODE REVIEW

**Comment by Subv** [comment]
Could you please include a hwtest for this?
Also, in the future, please don't include unrelated changes in the same PR, they make it much harder to review.

**Comment by fincs** [comment]
The test case from last night which resulted in this PR: https://feos.mtheall.com/citra-logicopbug.zip (originally intended just for the logic op bug, ended up discovering the rounding bug too)

**Comment by FearlessTobi** [comment]
https://github.com/citra-emu/citra/issues/2440 This should likely be fixed by skipping the null textures. I will test it later.

File: src/video_core/swrasterizer/rasterizer.cpp [hunk 1]
```diff
@@ -293,16 +293,16 @@ static void ProcessTriangleInternal(const Vertex& v0, const Vertex& v1, const Ve
             };
 
             Math::Vec4<u8> primary_color{
-                (u8)(
+                (u8)round(
```
Comment (B3n30) [inline src/video_core/swrasterizer/rasterizer.cpp]: is there a reason static_cast<u8>() isn't used?

File: src/video_core/swrasterizer/rasterizer.cpp [hunk 2]
```diff
@@ -331,6 +331,7 @@ static void ProcessTriangleInternal(const Vertex& v0, const Vertex& v1, const Ve
                 // Only unit 0 respects the texturing type (according to 3DBrew)
                 // TODO: Refactor so cubemaps and shadowmaps can be handled
                 PAddr texture_address = texture.config.GetPhysicalAddress();
+                if (texture_address == 0) continue;
```
Comment (B3n30) [inline src/video_core/swrasterizer/rasterizer.cpp]: Should there be a log if that case happens? Maybe debug

**Comment by B3n30** [comment]
What's the status of that PR? It needs clang-format fixing. Also my comments weren't addressed yet.

**Comment by ds84182** [comment]
I'll try to get back to this today but I discovered that this fixes Pokemon X&Y because textures with a null address get skipped with one of my changes. I did some GPU research and I concluded that the color output from a null address texture depends on the previous texture used and it varies if the types of the texture are different. The color is deterministic since it depends on what was drawn before it, but I'd conclude that **this is a bug in Pokemon X & Y**. The only reason why it wasn't fixed was because the developers _could not_ see what was happening with the null texture, since they always draw the male character's body before the two cheeks. In my tests I get slightly non-deterministic results when opening the home menu while using the null address. If a game depends on this weird hardware behavior for correct rendering we'd be in trouble though.

Anyways the correct thing to do is to warn in the log when a null texture is used and emit a transparent black. I've already argued that using the null texture is a bug in the game and is undefined (could break if Nintendo ever decides to do a new GPU revision, although I don't think they will), and we should provide a sensible default.

I'll remove the null address check out of software rasterizer for now, and I'll address the comments.

**Comment by MakerMayker** [comment]
I hope this works.

**Pull request merged**